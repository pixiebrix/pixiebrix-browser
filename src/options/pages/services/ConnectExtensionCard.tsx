import React, { useState } from "react";
import { useAsyncState } from "@/hooks/common";
import { getBaseURL } from "@/services/baseService";
import useAsyncEffect from "use-async-effect";
import { getExtensionToken } from "@/auth/token";
import Card from "react-bootstrap/Card";
import urljoin from "url-join";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faExternalLinkAlt } from "@fortawesome/free-solid-svg-icons";

const ConnectExtensionCard: React.FunctionComponent = ({}) => {
  const [hasExtensionKey, setHasExtensionKey] = useState(true);
  const [serviceURL] = useAsyncState(getBaseURL);

  useAsyncEffect(
    async (isMounted) => {
      const hasKey = !!(await getExtensionToken());
      if (isMounted()) return;
      setHasExtensionKey(hasKey);
    },
    [setHasExtensionKey]
  );

  if (hasExtensionKey) {
    return null;
  }

  return (
    <Card className="mb-4">
      <Card.Header>Connect to PixieBrix</Card.Header>
      <Card.Body>
        <p>
          By linking your browser extension with PixieBrix, you'll get access to
          team features and public services.
        </p>
        {serviceURL && (
          <a
            href={urljoin(serviceURL, "extension")}
            className="btn btn-primary"
            target="_blank"
          >
            <FontAwesomeIcon icon={faExternalLinkAlt} /> Open PixieBrix Website
          </a>
        )}
      </Card.Body>
    </Card>
  );
};

export default ConnectExtensionCard;
