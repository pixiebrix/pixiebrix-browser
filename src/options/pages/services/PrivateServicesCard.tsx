/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { useSelector } from "react-redux";
import { Button, Card, Table } from "react-bootstrap";
import React, { useCallback, useContext } from "react";
import { IService, RawServiceConfiguration } from "@/core";
import { RootState } from "@/options/store";
import { uuidv4 } from "@/types/helpers";
import BrickModal from "@/components/brickModal/BrickModal";
import {
  faEyeSlash,
  faPlus,
  faSignOutAlt,
} from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import AuthContext from "@/auth/AuthContext";
import { deleteCachedAuth } from "@/background/requests";
import { ServicesState } from "@/options/slices";
import useNotifications from "@/hooks/useNotifications";
import styles from "./PrivateServicesCard.module.scss";

const selectConfiguredServices = ({ services }: { services: ServicesState }) =>
  Object.values(services.configured);

type OwnProps = {
  services: IService[];
  navigate: (url: string) => void;
  onCreate: (configuration: RawServiceConfiguration) => void;
};

const PrivateServicesCard: React.FunctionComponent<OwnProps> = ({
  services,
  navigate,
  onCreate,
}) => {
  const notify = useNotifications();
  const { isLoggedIn } = useContext(AuthContext);

  const configuredServices = useSelector<RootState, RawServiceConfiguration[]>(
    selectConfiguredServices
  );

  const resetAuth = useCallback(
    async (authId: string) => {
      try {
        await deleteCachedAuth(authId);
        notify.success("Reset login for integration");
      } catch (error: unknown) {
        notify.error("Error resetting login for integration", {
          error,
        });
      }
    },
    [notify]
  );

  const onSelect = useCallback(
    (service: IService) => {
      onCreate({
        id: uuidv4(),
        label: undefined,
        serviceId: service.id,
        config: {},
      } as RawServiceConfiguration);
    },
    [onCreate]
  );

  return (
    <>
      <Card.Body className="pb-2 px-3">
        <p className="text-info">
          <FontAwesomeIcon icon={faEyeSlash} /> Private configurations are
          stored in your browser. They are never transmitted to the PixieBrix
          servers or shared with your team
        </p>
      </Card.Body>
      <Table responsive>
        <thead>
          <tr>
            <th>Label</th>
            <th>Type</th>
            <th>Actions</th>
          </tr>
        </thead>
        <tbody>
          {isLoggedIn && (
            <tr>
              <td>
                Zapier <i>&ndash; use to connect to PixieBrix from Zapier</i>
              </td>
              <td className="text-muted small">N/A</td>
              <td>
                <Button
                  style={{ width: 100 }}
                  variant="info"
                  size="sm"
                  onClick={() => {
                    navigate("/services/zapier/");
                  }}
                >
                  View Key
                </Button>
              </td>
            </tr>
          )}

          {configuredServices.map((configuredService) => {
            const service = services.find(
              (x) => x.id === configuredService.serviceId
            );
            if (!service) {
              throw new Error(`Unknown service ${configuredService.serviceId}`);
            }

            return (
              <tr
                key={`${configuredService.serviceId}-${configuredService.id}`}
              >
                <td>
                  {configuredService.label ? (
                    <span>{configuredService.label}</span>
                  ) : (
                    <span className="text-muted">No label provided</span>
                  )}
                </td>
                <td>
                  <div>{service.name}</div>
                  <div>
                    <code className="p-0" style={{ fontSize: "0.7rem" }}>
                      {service.id}
                    </code>
                  </div>
                </td>
                <td>
                  <Button
                    style={{ width: 100 }}
                    variant="info"
                    size="sm"
                    onClick={() => {
                      navigate(
                        `/services/${encodeURIComponent(configuredService.id)}`
                      );
                    }}
                  >
                    Configure
                  </Button>

                  {service.isOAuth2 || service.isToken ? (
                    <Button
                      size="sm"
                      variant="dark"
                      onClick={async () => resetAuth(configuredService.id)}
                    >
                      <FontAwesomeIcon icon={faSignOutAlt} /> Reset Token
                    </Button>
                  ) : (
                    <Button size="sm" variant="outline-dark" disabled>
                      <FontAwesomeIcon icon={faSignOutAlt} /> Reset Token
                    </Button>
                  )}
                </td>
              </tr>
            );
          })}
        </tbody>
      </Table>
      <Card.Footer>
        <BrickModal
          onSelect={onSelect}
          bricks={services}
          modalClassName={styles.ModalOverride}
          selectCaption={
            <span>
              <FontAwesomeIcon icon={faPlus} className="mr-1" /> Configure
            </span>
          }
          caption="Add Private Integration"
        />
      </Card.Footer>
    </>
  );
};

export default PrivateServicesCard;
