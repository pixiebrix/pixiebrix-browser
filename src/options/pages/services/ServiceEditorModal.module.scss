/*!
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

.modal {
  // 990 because the Google File Picker iframe is 1000; backdrop is 980 (see below)
  z-index: 990 !important;
}

.backdrop {
  // Modal is 980
  z-index: 980 !important;
}
