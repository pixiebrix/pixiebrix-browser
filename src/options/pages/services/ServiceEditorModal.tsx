/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import styles from "./ServiceEditorModal.module.scss";

import optionsRegistry from "@/components/fields/optionsRegistry";
import React, { useCallback, useMemo } from "react";
import { Modal, Button } from "react-bootstrap";
import AsyncButton from "@/components/AsyncButton";
import { type IService, type RawServiceConfiguration, type UUID } from "@/core";
import { dereference } from "@/validators/generic";
import { cloneDeep, truncate } from "lodash";
import { useAsyncState } from "@/hooks/common";
import genericOptionsFactory from "@/components/fields/schemaFields/genericOptionsFactory";
import { buildYup } from "schema-to-yup";
import * as Yup from "yup";
import reportError from "@/telemetry/reportError";
import { useTitle } from "@/hooks/title";
import ConnectedFieldTemplate from "@/components/form/ConnectedFieldTemplate";
import FieldTemplate from "@/components/form/FieldTemplate";
import FieldRuntimeContext, {
  type RuntimeContext,
} from "@/components/fields/schemaFields/FieldRuntimeContext";
import { OPTIONS_DEFAULT_RUNTIME_API_VERSION } from "@/options/constants";
import Form, {
  type OnSubmit,
  type RenderBody,
  type RenderSubmit,
} from "@/components/form/Form";

export type OwnProps = {
  configuration: RawServiceConfiguration;
  service: IService;
  onClose: () => void;
  onDelete?: (id: UUID) => void;
  onSave: (config: RawServiceConfiguration) => Promise<void>;
};

const FORM_RUNTIME_CONTEXT: RuntimeContext = {
  apiVersion: OPTIONS_DEFAULT_RUNTIME_API_VERSION,
  allowExpressions: false,
};

const ServiceEditorModal: React.FunctionComponent<OwnProps> = ({
  configuration: originalConfiguration,
  service,
  onClose,
  onDelete,
  onSave,
}) => {
  useTitle(`Configure ${truncate(service.name, { length: 15 })}`);

  const onSubmit = useCallback<OnSubmit<RawServiceConfiguration>>(
    async (values, helpers) => {
      helpers.setSubmitting(true);
      onClose();
      await onSave(values);
    },
    [onSave, onClose]
  );

  const Editor = useMemo(() => {
    if (optionsRegistry.has(service.id)) {
      return optionsRegistry.get(service.id);
    }

    return genericOptionsFactory(service.schema);
  }, [service]);

  const schemaPromise = useMemo(
    async () =>
      dereference({
        type: "object",
        properties: {
          organization: {
            type: "string",
          },
          label: {
            type: "string",
            // @ts-expect-error -- expects JSONSchema7 type `required: string[]`
            // (one level up), but only works with JSONSchema4 `required: boolean`
            required: true,
          },
          // $RefParse mutates the schema
          config: cloneDeep(service.schema),
        },
      }),
    [service.schema]
  );

  const [schema] = useAsyncState(schemaPromise);

  const validationSchema = useMemo(() => {
    if (!schema) {
      return Yup.object();
    }

    try {
      // The de-referenced schema is frozen, buildYup can mutate it, so we need to "unfreeze" the schema
      return buildYup(cloneDeep(schema), {});
    } catch (error) {
      console.error("Error building Yup validator from JSON Schema", { error });
      reportError(error, { logToConsole: false });
      return Yup.object();
    }
  }, [schema]);

  if (!schema) {
    return null;
  }

  const renderBody: RenderBody = () => (
    <Modal.Body>
      <FieldRuntimeContext.Provider value={FORM_RUNTIME_CONTEXT}>
        <ConnectedFieldTemplate
          name="label"
          label="Label"
          description="A label to help identify this integration"
          blankValue=""
        />
        <FieldTemplate
          label="Integration"
          name="service"
          type="text"
          plaintext
          readOnly
          value={service.id}
        />
        <Editor name="config" />
      </FieldRuntimeContext.Provider>
    </Modal.Body>
  );

  const renderSubmit: RenderSubmit = ({ isSubmitting, isValid }) => (
    <Modal.Footer>
      <div className="d-flex w-100">
        <div className="flex-grow-1">
          {onDelete && (
            <AsyncButton
              variant="outline-danger"
              onClick={() => {
                onDelete(originalConfiguration.id);
              }}
            >
              Delete
            </AsyncButton>
          )}
        </div>
        <div>
          <Button variant="default" onClick={onClose}>
            Cancel
          </Button>
          <Button
            variant="primary"
            type="submit"
            disabled={isSubmitting || !isValid}
          >
            Save
          </Button>
        </div>
      </div>
    </Modal.Footer>
  );

  return (
    <Modal
      show
      dialogClassName={styles.dialog}
      onHide={onClose}
      backdrop="static"
      keyboard={false}
    >
      <Modal.Header closeButton>
        <Modal.Title>Configure Private Integration: {service.name}</Modal.Title>
      </Modal.Header>

      <Form
        validationSchema={validationSchema}
        initialValues={originalConfiguration}
        onSubmit={onSubmit}
        renderBody={renderBody}
        renderSubmit={renderSubmit}
      />
    </Modal>
  );
};

export default ServiceEditorModal;
