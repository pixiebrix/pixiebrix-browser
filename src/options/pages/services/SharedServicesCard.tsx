/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { Card, Table } from "react-bootstrap";
import React, { useMemo, useState } from "react";
import { useAsyncState } from "@/hooks/common";
import { getBaseURL } from "@/services/baseService";
import urljoin from "url-join";
import GridLoader from "react-spinners/GridLoader";
import { SanitizedAuth } from "@/types/contract";
import { faUsers } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import Pagination from "@/components/pagination/Pagination";

interface OwnProps {
  remoteAuths: SanitizedAuth[];
}

const SERVICES_PER_PAGE = 10;

const SharedServicesCard: React.FunctionComponent<OwnProps> = ({
  remoteAuths,
}) => {
  const [serviceUrl] = useAsyncState(getBaseURL);
  const [page, setPage] = useState(0);

  const numPages = useMemo(
    () => Math.ceil(remoteAuths.length / SERVICES_PER_PAGE),
    [remoteAuths, SERVICES_PER_PAGE]
  );

  const pageServices = useMemo(
    () =>
      remoteAuths.slice(
        page * SERVICES_PER_PAGE,
        (page + 1) * SERVICES_PER_PAGE
      ),
    [remoteAuths, page]
  );

  return (
    <>
      <Card.Body className="pb-2 px-3">
        <p className="text-info">
          <FontAwesomeIcon icon={faUsers} /> Shared integrations made available
          by your team and/or built-in to PixieBrix
        </p>
      </Card.Body>
      {remoteAuths ? (
        <Table responsive>
          <thead>
            <tr>
              <th>Team</th>
              <th>Label</th>
              <th>Type</th>
            </tr>
          </thead>
          <tbody>
            {pageServices.map((remoteAuth) => (
              <tr key={remoteAuth.id}>
                <td>{remoteAuth.organization?.name ?? "âœ¨ Built-in"}</td>
                <td>{remoteAuth.label}</td>
                <td>
                  <div>{remoteAuth.service.config.metadata.name}</div>
                  <div>
                    <code className="p-0" style={{ fontSize: "0.7rem" }}>
                      {remoteAuth.service.config.metadata.id}
                    </code>
                  </div>
                </td>
              </tr>
            ))}
          </tbody>
        </Table>
      ) : (
        <Card.Body>
          <GridLoader />
        </Card.Body>
      )}
      <Card.Footer className="d-flex justify-content-between align-items-center">
        <span className="py-3">
          To configure shared services, open the{" "}
          <a
            href={urljoin(serviceUrl, "services")}
            target="_blank"
            rel="noopener noreferrer"
          >
            PixieBrix website
          </a>
        </span>
        <span className="text-muted">
          Showing {page * SERVICES_PER_PAGE + 1} to{" "}
          {SERVICES_PER_PAGE * page + pageServices.length} of{" "}
          {remoteAuths.length} integrations
        </span>
        {remoteAuths.length > SERVICES_PER_PAGE && (
          <Pagination page={page} setPage={setPage} numPages={numPages} />
        )}
      </Card.Footer>
    </>
  );
};

export default SharedServicesCard;
