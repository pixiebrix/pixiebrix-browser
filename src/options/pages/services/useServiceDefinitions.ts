/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { useSelector } from "react-redux";
import { RootState } from "@/options/store";
import { IService, RawServiceConfiguration } from "@/core";
import { useParams } from "react-router";
import { useMemo } from "react";
import { sortBy } from "lodash";
import registry from "@/services/registry";
import { useAsyncState } from "@/hooks/common";
import { ServicesState } from "@/options/slices";
import { PIXIEBRIX_SERVICE_ID } from "@/services/constants";

interface ServiceDefinitions {
  serviceDefinitions: IService[];
  activeConfiguration: RawServiceConfiguration | null;
  activeService: IService | null;
  isPending: boolean;
  showZapier: boolean;
}

const ZAPIER_SLUG = "zapier";

const selectConfiguredServices = ({ services }: { services: ServicesState }) =>
  Object.values(services.configured);

function useServiceDefinitions(): ServiceDefinitions {
  const configuredServices = useSelector<RootState, RawServiceConfiguration[]>(
    selectConfiguredServices
  );
  const { id: configurationId } = useParams<{ id: string }>();

  const showZapier = configurationId === ZAPIER_SLUG;

  const [serviceDefinitions, isPending, error] = useAsyncState(
    async () =>
      sortBy(
        (await registry.all()).filter((x) => x.id !== PIXIEBRIX_SERVICE_ID),
        (x) => x.id
      ),
    []
  );

  const activeConfiguration = useMemo(
    () =>
      configurationId && configurationId !== ZAPIER_SLUG
        ? configuredServices.find((x) => x.id === configurationId)
        : null,
    [configuredServices, configurationId]
  );

  const activeService = useMemo(
    () =>
      activeConfiguration
        ? (serviceDefinitions ?? []).find(
            (x) => x.id === activeConfiguration.serviceId
          )
        : null,
    [serviceDefinitions, activeConfiguration]
  );

  if (error) {
    throw error;
  }

  return {
    activeConfiguration,
    serviceDefinitions,
    activeService,
    isPending,
    showZapier,
  };
}

export default useServiceDefinitions;
