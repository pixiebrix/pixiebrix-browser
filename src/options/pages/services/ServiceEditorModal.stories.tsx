/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { type ComponentProps } from "react";
import { type ComponentMeta, type Story } from "@storybook/react";
import ServiceEditorModal from "./ServiceEditorModal";
import { action } from "@storybook/addon-actions";
import { fromJS } from "@/services/factory";
import { dumpBrickYaml, loadBrickYaml } from "@/runtime/brickYaml";
import { type ServiceDefinition } from "@/types/definitions";

import pipedriveYaml from "@contrib/services/pipedrive.yaml?loadAsText";
import automationAnywhereYaml from "@contrib/services/automation-anywhere.yaml?loadAsText";
import registerDefaultWidgets from "@/components/fields/schemaFields/widgets/registerDefaultWidgets";
import { settingsStore } from "@/testUtils/storyUtils";
import { Provider } from "react-redux";

const FIXTURES = {
  pipedrive: pipedriveYaml,
  automationAnywhere: automationAnywhereYaml,
};

type StoryType = ComponentProps<typeof ServiceEditorModal> & {
  fixture: keyof typeof FIXTURES;
};

export default {
  title: "Options/ServiceEditorModal",
  component: ServiceEditorModal,
  // Modals are not compatible with Storyshots
  parameters: {
    storyshots: false,
  },
} as ComponentMeta<typeof ServiceEditorModal>;

const Template: Story<StoryType> = ({ fixture, ...args }) => {
  const service = fromJS(
    // eslint-disable-next-line security/detect-object-injection -- type checked from fixture object
    loadBrickYaml(dumpBrickYaml(FIXTURES[fixture])) as ServiceDefinition
  );

  // Cheap call, just call in the render function
  registerDefaultWidgets();

  return (
    <Provider store={settingsStore()}>
      <ServiceEditorModal {...args} service={service} />
    </Provider>
  );
};

export const Pipedrive = Template.bind({});
Pipedrive.args = {
  fixture: "pipedrive",
  onDelete: action("onDelete"),
  onClose: action("onClose"),
  async onSave() {
    action("onSave");
  },
};

export const AutomationAnywhere = Template.bind({});
AutomationAnywhere.args = {
  fixture: "automationAnywhere",
  onDelete: action("onDelete"),
  onClose: action("onClose"),
  async onSave() {
    action("onSave");
  },
};
