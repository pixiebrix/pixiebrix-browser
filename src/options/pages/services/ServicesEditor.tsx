/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React, { useCallback, useContext, useState } from "react";
import { connect, useSelector } from "react-redux";
import { servicesSlice, ServicesState } from "../../slices";
import { PageTitle } from "@/layout/Page";
import { Row, Col, Card, Nav, Badge } from "react-bootstrap";
import { push } from "connected-react-router";
import { useToasts } from "react-toast-notifications";
import ServiceEditorModal from "./ServiceEditorModal";
import PrivateServicesCard from "./PrivateServicesCard";
import ConnectExtensionCard from "./ConnectExtensionCard";
import SharedServicesCard from "./SharedServicesCard";
import { faCloud } from "@fortawesome/free-solid-svg-icons";
import { useFetch } from "@/hooks/fetch";
import useServiceDefinitions from "./useServiceDefinitions";
import { RootState } from "../../store";
import { RawServiceConfiguration } from "@/core";
import { SanitizedAuth } from "@/types/contract";
import { refresh as refreshServices } from "@/background/locator";
import GridLoader from "react-spinners/GridLoader";
import ZapierModal from "@/options/pages/services/ZapierModal";
import AuthContext from "@/auth/AuthContext";
import { sleep } from "@/utils";
import { reportError } from "@/telemetry/logging";
import { useTitle } from "@/hooks/title";

const { updateServiceConfig, deleteServiceConfig } = servicesSlice.actions;

interface OwnProps {
  updateServiceConfig: (config: RawServiceConfiguration) => void;
  deleteServiceConfig: ({ id }: { id: string }) => void;
  navigate: (url: string) => void;
}

const selectConfiguredServices = ({ services }: { services: ServicesState }) =>
  Object.values(services.configured);

const ServicesEditor: React.FunctionComponent<OwnProps> = ({
  updateServiceConfig,
  deleteServiceConfig,
  navigate,
}) => {
  useTitle("Integrations");

  const remoteAuths = useFetch<SanitizedAuth[]>("/api/services/shared/?meta=1");
  const { flags } = useContext(AuthContext);

  const configuredServices = useSelector<RootState, RawServiceConfiguration[]>(
    selectConfiguredServices
  );

  const [activeTab, setTab] = useState("private");

  const { addToast } = useToasts();

  const {
    activeConfiguration,
    serviceDefinitions,
    activeService,
    showZapier,
    isPending: servicesPending,
  } = useServiceDefinitions();

  const handleSave = useCallback(
    async (config) => {
      updateServiceConfig(config);
      addToast(`Updated your private configuration for ${activeService.name}`, {
        appearance: "success",
        autoDismiss: true,
      });

      // wait 1000 to allow changes to propagate to storage
      sleep(1000).then(async () => {
        try {
          await refreshServices();
        } catch (error) {
          reportError(error);
          addToast(
            `Error refreshing service configurations, restart the PixieBrix extension`,
            {
              appearance: "warning",
              autoDismiss: true,
            }
          );
        }
      });

      navigate("/services");
    },
    [updateServiceConfig, activeService, addToast, navigate]
  );

  const handleDelete = useCallback(
    async (id) => {
      deleteServiceConfig({ id });

      addToast(`Deleted private configuration for ${activeService.name}`, {
        appearance: "success",
        autoDismiss: true,
      });

      sleep(1000).then(async () => {
        try {
          await refreshServices();
        } catch (error) {
          addToast(
            `Error refreshing service configurations, restart the PixieBrix extension`,
            {
              appearance: "warning",
              autoDismiss: true,
            }
          );
          reportError(error);
        }
      });

      navigate("/services");
    },
    [deleteServiceConfig, navigate, addToast, activeService]
  );

  if (servicesPending) {
    return (
      <div>
        <PageTitle icon={faCloud} title="Integrations" />
        <div className="pb-4">
          <p>
            Configure external accounts, resources, and APIs that you can re-use
            across bricks
          </p>
        </div>
        <Row>
          <Col>
            <GridLoader />
          </Col>
        </Row>
      </div>
    );
  }

  return (
    <div>
      <PageTitle icon={faCloud} title="Integrations" />
      <div className="pb-4">
        <p>
          Configure external accounts, resources, and APIs that you can re-use
          across bricks
        </p>
      </div>
      {showZapier && <ZapierModal onClose={() => navigate("/services")} />}
      {activeConfiguration && activeService && (
        <ServiceEditorModal
          configuration={activeConfiguration}
          service={activeService}
          onDelete={handleDelete}
          onClose={() => navigate("/services")}
          onSave={handleSave}
        />
      )}
      <Row>
        <Col>
          <ConnectExtensionCard />
        </Col>
      </Row>
      <Row>
        <Col>
          <Card>
            <Card.Header>
              <Nav
                variant="tabs"
                defaultActiveKey={activeTab}
                onSelect={(x: string) => setTab(x)}
              >
                <Nav.Item>
                  <Nav.Link eventKey="private">
                    Private Integrations{" "}
                    <Badge variant="info">
                      {configuredServices ? configuredServices.length : "?"}
                    </Badge>
                  </Nav.Link>
                </Nav.Item>
                {flags.includes("teams") && (
                  <Nav.Item>
                    <Nav.Link eventKey="shared">
                      Shared Integrations{" "}
                      <Badge variant="info">
                        {remoteAuths ? remoteAuths.length : "?"}
                      </Badge>
                    </Nav.Link>
                  </Nav.Item>
                )}
              </Nav>
            </Card.Header>

            {activeTab === "private" && (
              <PrivateServicesCard
                navigate={navigate}
                services={serviceDefinitions}
                onCreate={(config) => {
                  updateServiceConfig(config);
                  navigate(`/services/${encodeURIComponent(config.id)}`);
                }}
              />
            )}

            {flags.includes("teams") && activeTab === "shared" && (
              <SharedServicesCard remoteAuths={remoteAuths} />
            )}
          </Card>
        </Col>
      </Row>
    </div>
  );
};

export default connect(null, {
  updateServiceConfig,
  deleteServiceConfig,
  navigate: push,
})(ServicesEditor);
