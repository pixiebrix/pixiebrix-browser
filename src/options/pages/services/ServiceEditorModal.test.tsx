/*
 * Copyright (C) 2022 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React from "react";
import { fromJS } from "@/services/factory";
import ServiceEditorModal from "@/options/pages/services/ServiceEditorModal";
import { render, screen } from "@testing-library/react";
import { waitForEffect } from "@/tests/testHelpers";

// FIXME: this is coming through as a module with default being a JSON object. (yaml-jest-transform is being applied)
import pipedriveYaml from "@contrib/services/pipedrive.yaml?loadAsText";
import { RawServiceConfiguration } from "@/core";

describe("ServiceEditorModal", () => {
  test("Can render Pipedrive configuration modal without existing configuration", async () => {
    const service = fromJS(pipedriveYaml as any);

    const rendered = render(
      <ServiceEditorModal
        configuration={{ label: "" } as RawServiceConfiguration}
        onDelete={jest.fn()}
        onSave={jest.fn()}
        onClose={jest.fn()}
        service={service}
      />
    );

    await waitForEffect();

    expect(screen.getByDisplayValue("pipedrive/api")).not.toBeNull();
    expect(screen.getByText("Save")).not.toBeNull();
    expect(screen.getByText("Delete")).not.toBeNull();
    expect(screen.getByText("Close")).not.toBeNull();

    expect(rendered.asFragment()).toMatchSnapshot();
  });
});
