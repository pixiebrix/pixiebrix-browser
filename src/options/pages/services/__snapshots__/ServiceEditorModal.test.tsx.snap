// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceEditorModal Can render Pipedrive configuration modal without existing configuration 1`] = `<DocumentFragment />`;
